sklearn/inspection/_permutation_importance.py

```diff
@@ -104,20 +106,23 @@ def permutation_importance(estimator, X, y, scoring=None, n_repeats=5,
     .. [BRE] L. Breiman, "Random Forests", Machine Learning, 45(1), 5-32,
              2001. https://doi.org/10.1023/A:1010933404324
     """
-    if hasattr(X, "iloc"):
-        X = X.copy()  # Dataframe
-    else:
-        X = check_array(X, force_all_finite='allow-nan', dtype=np.object,
-                        copy=True)
-
+    if not hasattr(X, "iloc"):
+        X = check_array(X, force_all_finite='allow-nan', dtype=None)
```

Comment:
Notes for reviewers:

- we do not need to copy anything here now: the full X array copy is now delayed to the `_calculate_permutation_scores` function itself to ensure thread-safety.
- I have no idea why there was `dtype=np.object` before. This can make the size of X blow up and furthermore it would have prevent memory mapping in joblib (which arguable has limited use because we make a copy of X in the worker processes anyway, still memmaping can maybe speed-up interprocess communication a bit).

Reply:
So do we need the fix in #15898?

Reply:
No,  #15898 is no longer needed since the current PR always does a copy of X inside `_calculate_permutation_scores`. Even if X is backed by a readonly memory buffer, then `X.copy()` will always be backed by a writable memory buffer.

The current PR also includes a non regression test for #15810, see: `test_permutation_importance_large_memmaped_data` below.